File: sally/leetcode/128_ Longest Consecutive Sequence.cpp
// 128. Longest Consecutive Sequence
// Runtime: 57 ms, faster than 85.98% of C++ online submissions for Longest Consecutive Sequence.
// Memory Usage: 22.3 MB, less than 97.56% of C++ online submissions for Longest Consecutive Sequence.

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        nums.erase(unique(nums.begin(), nums.end()), nums.end());
        int n_len = nums.size();
        int result = 1;
        int max_len = 1;
        for(int i = 1; i < n_len; i++){
            if(nums[i-1]+1 == nums[i]) max_len++;
            else {
                result = max(max_len, result);
                max_len = 1;
            }
        }
        result = max(max_len, result);
        if(n_len == 0) result = 0;
        return result;
    }
};

File: coodingpenguin/leetcode/128_LongestConsecutiveSequence.py
# 문제: 128. Longest Consecutive Sequence
# 링크: https://leetcode.com/problems/longest-consecutive-sequence/


# 시간/공간: 237ms / 25.9MB
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # 길이가 1이하라면
        if len(nums) <= 1:
            return len(nums)

        count = 1  # 연속 시퀀스 길이
        max_length = 1  # 최대 연속 시퀀스 길이
        sorted_nums = sorted(set(nums))  # 고유한 요소에 대해 정렬

        for i in range(1, len(sorted_nums)):
            # 연속 시퀀스를 이룬다면
            if sorted_nums[i - 1] == sorted_nums[i] - 1:
                count += 1  # 길이 갱신
            # 아니라면
            else:
                max_length = max(max_length, count)  # 최대 길이 갱신
                count = 1  # 길이 초기화
        max_length = max(max_length, count)  # 마지막 요소에 대한 갱신
        return max_length
